### ISSUE TITLE
KV cache and autoregressive decoding

### ISSUE THREAD

**Issue by anibjoshi** [strata-ai-labs/strata-inference#16]
### Context
Beyond embedding, strata-inference should support text generation (completion, chat). This requires:
1. **KV cache**: Store computed key/value tensors to avoid recomputation
2. **Causal attention mask**: Tokens only attend to previous positions
3. **Autoregressive loop**: Generate one token at a time, feeding it back

### KV Cache Design
```rust
pub struct KvCache {
    /// One entry per layer
    layers: Vec<KvCacheLayer>,
    /// Current sequence length (grows with each generated token)
    seq_len: usize,
    /// Maximum sequence length
    max_len: usize,
}

pub struct KvCacheLayer {
    /// [cached_seq_len, num_kv_heads * head_dim]
    k: DeviceTensor,
    /// [cached_seq_len, num_kv_heads * head_dim]
    v: DeviceTensor,
}
```

### Generation Loop
```
1. Prompt processing (prefill):
   - Tokenize full prompt
   - Run through all layers, caching K/V at each layer
   - Get logits for last token position
   
2. Autoregressive decode (one token at a time):
   - Apply sampling to logits → next token
   - If stop condition (EOS, max_len) → done
   - Run single new token through all layers:
     - Compute Q from new token only
     - Compute K, V from new token, append to cache
     - Attention: Q (1 token) × K (all cached) → scores
     - Attention: scores × V (all cached) → output
   - Get logits → repeat
```

### Sampling Strategies
```rust
pub enum SamplingStrategy {
    Greedy,
    TopK { k: usize },
    TopP { p: f32 },
    Temperature { temp: f32 },
    Combined { temp: f32, top_k: usize, top_p: f32 },
}
```

### Acceptance Criteria
- [ ] KV cache allocation and management
- [ ] Causal attention mask generation
- [ ] Prefill: process full prompt in one forward pass
- [ ] Decode: process single token with cached KV
- [ ] Greedy sampling (argmax)
- [ ] Top-k sampling
- [ ] Top-p (nucleus) sampling
- [ ] Temperature scaling
- [ ] Stop conditions: EOS token, max length, stop strings
- [ ] Streaming output (yield tokens as generated)
- [ ] Test with a small LLaMA/Gemma model

### API Design
```rust
pub struct GenerationEngine {
    model: Model,
    tokenizer: Box<dyn Tokenizer>,
    config: ModelConfig,
    backend: Arc<dyn ComputeBackend>,
}

impl GenerationEngine {
    pub fn generate(&self, prompt: &str, params: GenerationParams) -> Result<String>;
    pub fn generate_stream(&self, prompt: &str, params: GenerationParams) -> impl Iterator<Item = String>;
}
```

### Estimated LOC: ~1,200-1,500
